src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -118,28 +122,29 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
 
         using (Logger.LogBlock(FunctionId.AssetService_SynchronizeAssetsAsync, Checksum.GetChecksumsLogInfo, checksums, cancellationToken))
         {
-            using var _1 = ArrayBuilder<Checksum>.GetInstance(checksums.Count, out var missingChecksums);
+            using var missingChecksumsObject = s_checksumListPool.GetPooledObject();
```

Comment:
This intermediate collection is completely unnecessary. Just use the `checksums` argument directly:

```cs
// Remove all values from 'checksums' where the asset cache already contains a value
checksums.RemoveWhere(checksum => _assetCache.TryGetAsset<object>(checksum, out _));
```

This can be further optimized by defining `PooledDelegates.GetPooledPredicate`, if necessary:

```cs
// Remove all values from 'checksums' where the asset cache already contains a value
using (var _1 = PooledDelegates.GetPooledPredicate<Checksum, AssetProvider>(
    static (checksum, arg) => arg._assetCache.TryGetAsset<object>(checksum, out _),
    this,
    out var predicate))
{
    checksums.RemoveWhere(predicate);
}
```

Reply:
I don't think I'm following this suggestion. 

Modifying that hashset in place requires knowledge that the caller is ok with that. It might be ok, but I'd need to verify that. 

Assuming that were ok, then I'd need to pass that HashSet down and change all the calls to take in a HashSet instead of a SegmentedList, right?

Reply:
> Modifying that hashset in place requires knowledge that the caller is ok with that. It might be ok, but I'd need to verify that.

I did verify it, but we'd want a way to make it apparent to the callers.

> Assuming that were ok, then I'd need to pass that HashSet down and change all the calls to take in a HashSet instead of a SegmentedList, right?

That's one option. We could also create an `ImmutableArray` from it directly by calling `ImmutableArray.CreateRange<T>`, without going through the builder pool.

Reply:
The intent of this PR was to investigate removing the allocation of creating that ImmutableArray. If we *need* that data structure to be passed down as immutable, I should just close out this PR. But, from what I saw, it didn't need to be immutable, it was just a nice to have. This is an investigation into whether removing that nice to have could have a positive perf impact without too much ugliness in the code.

As for the suggestion to modify the HashSet in-place, I guess I'm still not seeing the value. Perhaps the speedometer profile with this change will show that's an issue, but I'd prefer to not make that change unless it's needed as it does change the semantics around the hashset passed into this method.

Reply:
as these are all internal components, without a surface area exposed to other features, i would be fine with this not being immutable, esp if it cuts down on churn.